### PR TITLE
proposal - Sedna controller enhancement

### PR DESCRIPTION
**Pull request by SherlockShemol**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

**Comment by SherlockShemol** [comment]
The pod recreate logic distinguishes whether the failed pod is a cloud worker or an edge worker, deletes the failed pod and recreates it continuously, and abandons the recreate if it still fails after 10 attempts.**The next step is to verify whether the solution is feasible.**

The following is the main logic:
filter failed pods.
```go
// filterFailedWorker filters the failed pods and returns the filtered failed pods, bool value of cloudWorkerFailed and edgeWorkerFailed
func filterFailedWorker(pods []*v1.Pod) (failedPods []*v1.Pod, cloudWorkerFailed bool, edgeWorkerFailed bool) {
    for _, pod := range pods {
        if pod.Status.Phase == v1.PodFailed {
            failedPods = append(failedPods, pod)
            if strings.Contains(pod.Name, "cloudworker") {
                cloudWorkerFailed = true
            } else if strings.Contains(pod.Name, "edgeworker") {
                edgeWorkerFailed = true
            }
        }
    }
    return failedPods, cloudWorkerFailed, edgeWorkerFailed
}
```
The method of distinguishing whether it is a cloud worker or an edge worker requires further confirmation.

get failed pods:
```go
    if failed > 0 {
        serviceFailed = true
        // TODO: get the failed worker, and knows that which worker fails, edge inference worker or cloud inference worker
        failedPods, cloudWorkerFailed, edgeWorkerFailed := filterFailedWorker(pods)
        if cloudWorkerFailed {
            reason = "cloudWorkerFailed"
            message = "the cloud worker of service failed"
        } else if edgeWorkerFailed {
            reason = "edgeWorkerFailed"
            message = "the edge worker of service failed"
        }
        newCondtionType = sednav1.JointInferenceServiceCondFailed
        c.recorder.Event(&service, v1.EventTypeWarning, reason, message)
```
delete and keep trying to recreate failed pods for 10 times or until success:
```go
        // delete and keep trying to recreate failed pods for 10 times or until success
        // delete pods one by one
        for _, pod := range failedPods {
            err = c.kubeClient.CoreV1().Pods(pod.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
            if err != nil {
                klog.Warningf("failed to delete pod: %v", err)
            } else {
                active--
            }
        }
        for {
            createCloudTempCount := 0
            createEdgeTempCount := 0
            if cloudWorkerFailed {
                err = c.createCloudWorker(&service, BigModelPort)
                createCloudTempCount++
                if err != nil {
                    klog.Warningf("failed to create cloud worker: %v", err)
                } else {
                    cloudWorkerFailed = false
                    active++
                    // escape the loop
                    break
                }
            }
            if edgeWorkerFailed {
                err = c.createEdgeWorker(&service, "localhost", BigModelPort)
                createEdgeTempCount++
                if err != nil {
                    klog.Warningf("failed to create edge worker: %v", err)
                } else {
                    edgeWorkerFailed = false
                    active++
                    // escape the loop
                    break
                }
            }
            // if failed to create cloud worker and edge worker for 10 times, Abandon pod recreate and escape the loop
            if createCloudTempCount == 10 || createEdgeTempCount == 10 {
                // escape the loop
                break
            }
        }
```

**Review by MooreZheng — APPROVED** [review]
This PR fixed https://github.com/kubeedge/sedna/issues/430
Previous comments can be found at https://github.com/kubeedge/sedna/pull/435

The concerns about compatibility are considered in this proposal. As mentioned in the scope, "Sedna joint inference and federated learning controller optimization only modifies the joint inference and federated learning controller codes. The controllers are decoupled from each other, and the function interface and library functions are not modified, so it will not affect the functions of other controllers. For existing application cases, since the CRD is not modified, the deployment of existing applications will not be affected."

The proposal looks good to me. The bug fix is not a trivial job. Sedna is also quite a complicated system. The proposal could be developed in a continuous manner and be modified during future implementations if needed.

File: pkg/globalmanager/controllers/jointinference/jointinferenceservice.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ const (
 )
 
 // Kind contains the schema.GroupVersionKind for this controller type.
-var Kind = sednav1.SchemeGroupVersion.WithKind(Name)
+var Kind = sednav1.SchemeGroupVersion.WithKind(KindName)
```
Comment (tangming1996) [inline pkg/globalmanager/controllers/jointinference/jointinferenceservice.go:66]: The bug fix and documentation PRs should be submitted separately. Please keep this in mind for future submissions.

Comment (SherlockShemol) [reply pkg/globalmanager/controllers/jointinference/jointinferenceservice.go:66]: Yes, I get it.Thank you, Mr. Tang.

**Comment by SherlockShemol** [comment]
> This PR fixed #430 Previous comments can be found at #435
> 
> The concerns about compatibility are considered in this proposal. As mentioned in the scope, "Sedna joint inference and federated learning controller optimization only modifies the joint inference and federated learning controller codes. The controllers are decoupled from each other, and the function interface and library functions are not modified, so it will not affect the functions of other controllers. For existing application cases, since the CRD is not modified, the deployment of existing applications will not be affected."
> 
> The proposal looks good to me. We also got another lgtm at the routine meeting. It would be fine to merge now. However, the bug fix is not a trivial job. Sedna is also quite a complicated system. The proposal could be developed in a continuous manner and be modified during future implementations if needed.

Thank you, Mr. Zheng.I will continue to improve the proposal and carefully implement the solution.

**Pull request merged**